docs/performance/inserts/bulk.txt

```diff
@@ -0,0 +1,147 @@
+.. _bulk_inserts:
+
+============
+Bulk Inserts
+============
+
+.. rubric:: Table of Contents
+
+.. contents::
+   :local:
+
+Single Inserts
+==============
+
+`Single inserts`_ use the ``INSERT`` statement, and look like this:
+
+.. code-block:: psql
+
+    cr> INSERT INTO my_table (column_a) VALUES ("value 1");
+    INSERT OK, 1 row affected (... sec)
+
+Single inserts are typically very fast with CrateDB. A small cluster can
+easily handle several thousand inserts per second.
+
+However, it's important to note that:
+
+ 1. Every insert is first applied to the primary shard
+
+ 2. After the primary shard has been updated, the insert is then individually
+    communicated in parallel to every configured replica shard
+
+ 3. CrateDB will not return a response until all replica shards have been
+    updated
+
+The overhead for each insert (query parsing, planning, and execution, which
+includes network connection setups, connection teardowns, and connection
+metadata) starts to add up for very heavy insert workloads.
+
+In addition, lots of internal traffic will congest your network, which will
+slow down any network-based cluster operations (i.e. inserts, distributed
+reads, replication, and cluster management).
+
+Fortunately, there is a solution: bulk inserts.
+
+Switching to Bulk Inserts
+=========================
+
+If you can batch up your inserts, switching to *bulk inserts* will dramatically
+improve both ingestion performance, and overall cluster performance.
+
+Bulk inserts make use of the CrateDB HTTP endpoint `bulk operations`_ feature
```

Comment:
Some postgres based clients also provide functionality which let us create/queue up a bulk request internally. (There are still individual messages from the client, but these are lower overhead because they're binary and contain almost no headers and happen on an already established connection).

Furthermore, bulk insert can also be accomplished by using multiple values clauses (though it's not *exactly* the same, as using the bulk operations features)

Reply:
@mfussenegger can you provide the names of the clients that offer bulk requests? we don't have to list all of them. but the the noteworthy ones would be good

also, re using multiple values clauses. I tried this and couldn't get it to work. I also couldn't see anything in the docs. can you provide an example query?

Reply:
I guess the noteworthy one is JDBC (`addBatch` and `executeBatch`),  as it's the only one where we also have tests for it.

multiple values: `insert into t (x, y) values (1, 1), (2, 2), (3, 3)`